dask/highlevelgraph.py

```diff
@@ -119,27 +119,43 @@ def from_collections(cls, name, layer, dependencies=()):
         ...     graph = HighLevelGraph.from_collections(name, layer, dependencies=[self])
         ...     return new_collection(name, graph)
         """
-        layers = {name: layer}
-        deps = {}
-        deps[name] = set()
-        for collection in toolz.unique(dependencies, key=id):
-            if is_dask_collection(collection):
-                graph = collection.__dask_graph__()
-                if isinstance(graph, HighLevelGraph):
-                    layers.update(graph.layers)
-                    deps.update(graph.dependencies)
-                    with ignoring(AttributeError):
-                        deps[name] |= set(collection.__dask_layers__())
-                else:
-                    try:
-                        [key] = collection.__dask_layers__()
-                    except AttributeError:
-                        key = id(graph)
-                    layers[key] = graph
-                    deps[name].add(key)
-                    deps[key] = set()
-            else:
-                raise TypeError(type(collection))
+        if any(toolz.remove(is_dask_collection, toolz.unique(dependencies, key=id))):
+            raise TypeError
+
+        graphs_and_collections = [
+            (collection.__dask_graph__(), collection)
+            for collection in toolz.unique(dependencies, key=id)
+        ]
+
+        hlgs = filter(
+            lambda x: isinstance(x[0], HighLevelGraph), graphs_and_collections
+        )
```

Comment:
In my experience list comprehensions are generally faster than `filter`.

```python
In [1]: data = list(range(10000))

In [2]: %timeit [i for i in data if i < 5000]
311 µs ± 3.08 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [3]: %timeit list(filter(lambda i: i < 5000, data))
751 µs ± 8.33 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Reply:
I think I need to learn more about iterators...